crypto-attributes/src/entity_derive.rs

```diff
@@ -0,0 +1,567 @@
+use proc_macro2::{Ident, Span};
+use quote::quote;
+use syn::spanned::Spanned;
+use syn::{Attribute, Data, DataStruct, Fields, FieldsNamed, Token, Type};
+
+/// Representation of a struct annotated with `#[derive(Entity)]`.
+pub(super) struct KeyStoreEntity {
+    /// Name of the type to implement the trait on
+    struct_name: Ident,
+    /// Database table name
+    collection_name: String,
+    /// The ID column
+    id: IdColumn,
+    /// All other columns
+    columns: Vec<Column>,
+    /// Whether to fail on inserting conflicting ids instead of using upsert semantics (default: false)
+    no_upsert: bool,
+}
+
+impl KeyStoreEntity {
+    pub(super) fn from_ast(input: &syn::DeriveInput) -> syn::Result<Self> {
+        let struct_name = input.ident.clone();
+
+        // #[entity(collection_name = "my_collection")]
+        let (collection_name, no_upsert) = Self::parse_collection_name_and_no_upsert(&input.attrs)?;
+        if collection_name.is_empty() {
+            return Err(syn::Error::new_spanned(
+                input,
+                format!(
+                    "Provide a collection in the `entity` attribute, e.g: `#[entity(collection_name = \"{}\")]`.",
+                    struct_name.to_string().to_lowercase()
+                ),
+            ));
+        }
+
+        let fields = Self::fields_from_data(&input.data, input.span())?;
+        let id = IdColumn::parse(fields, input.span())?;
+        let columns = Column::parse_columns(fields)?;
+
+        Ok(KeyStoreEntity {
+            struct_name,
+            collection_name,
+            id,
+            columns,
+            no_upsert,
+        })
+    }
+
+    /// Convert Keystore entity to a flattened version that is more verbose but can be used more easily in `quote!()`.
+    pub(super) fn flatten(self) -> KeyStoreEntityFlattened {
+        let all_columns = self.columns.iter().map(|column| column.name.clone()).collect();
+
+        let blob_columns = self
+            .columns
+            .iter()
+            .filter_map(|column| {
+                if matches!(column.column_type, ColumnType::Bytes) {
+                    Some(column.name.clone())
+                } else {
+                    None
+                }
```

Comment:
Could be

```rust
(column.column_type == ColumnType::Bytes).then(|| column.name.clone())
```

I don't expect the performance impact of deriving `PartialEq, Eq` for `ColumnType` to be severe.